### DISCUSSION TITLE
Is it possible to disable backpressure?

### DISCUSSION THREAD

**Discussion by seiyria** [uNetworking/uWebSockets.js#410] [General]
I'm making a game, and admittedly I don't get what backpressure is or how it works, but it seems like it will forcibly disconnect some clients which I don't want.

**Comment by ghost** [comment]
The connection is a queue where whatever you shove into it, must flow out of it (in order). So if you shove a lot of data into a connection that is slow or just refusing to consume the data, you will get backpressure (queue built-up). You don't have to disconnect, but you have to at least skip sending (which is what games do). So you really only need to do something like:

if (connection.getBufferedAmount() == 0) {
// do the actual send of new data here
}
// otherwise do nothing

**Reply by seiyria** [reply]
So in this case, what's the difference between `drain()` and `message()`? A message is something that comes through right away, but `drain()` is called when there are too many built up messages?

**Reply by ghost** [reply]
Only the sender can have backpressure, so "message" is receiving a message and "drain" is draining the (outgoing) queue (backpressure) by anywhere from 0 to all bytes. You don't need to care about "drain" if you do what I proposed above.

**Reply by seiyria** [reply]
Understood. Thanks! So, how exactly do I get the outgoing messages and send them? I'm looking at the examples, and the only one I found with a `drain()` is this one (which seems to do something random):

https://github.com/uNetworking/uWebSockets.js/blob/369b4a551ffed7f31d147c0c6246d3c8fa0c8e41/examples/Backpressure.js#L47-L53

It looks like I can get the buffered amount to check if there _is_ buffered content, but based on what I'm seeing in the API I can't tell how to let it send: 

![image](https://user-images.githubusercontent.com/763609/102514691-63281800-4052-11eb-8975-2f36f2cf6503.png)

**Comment by ghost** [comment]
I'm pretty sure you're overcomplicating things now. You don't really need to bother. The only thing you need to do is check buffered amount before sending that's all.

**Reply by seiyria** [reply]
I probably am, but I'm having a lot of trouble understanding the examples here. I think I understand the role of `drain`, but I don't quite get how to get the data _to_ drain. It doesn't seem to be a function where you return whether or not it can drain, so the example code you give:

```
if (connection.getBufferedAmount() == 0) {
   // do the actual send of new data here
}
```

My question really boils down to "how do I get the new data?"

**Reply by hst-m** [reply]
> how exactly do I get the outgoing messages and send them?
> My question really boils down to "how do I get the new data?"

all the messages you tried to send are in the queue and will be sent when they can be, you dont need to do anything to send them. the "drain" event is just called when any extra backpressure gets drained, you dont need to do anything there unless you want to

**Reply by seiyria** [reply]
I see, that helps a ton, thank you!

**Comment by ghost** [comment]
You actually don't even need to do that check I wrote. It is automatic if you set maxBackpressure: 512 or something. Then the library skips sending anything that can't fit in the queue. So it really is nothing you need to bother with.

**Reply by hst-m** [reply]
so you need to update docs that say is for publish only? https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocketbehavior.html#maxbackpressure

**Reply by hst-m** [reply]
looks like the ws.send() skip on maxBackpressure got added after the docs for maxBackpressure were added https://github.com/uNetworking/uWebSockets/commit/e85b5bff3c3197c256bb40965efc96a2de962311 can update the docs
